### PR TITLE
Create role to deploy MFE for native installations

### PR DESCRIPTION
**Pull request by morenol**
This PR creates three roles to deploy the profile, gradebook and account MFEs. 

This could be the starting point, the MFE are being served with nginx. We could modify this  in the future to allow the usage of S3 or any other service. I also think that if the files are going to be served directly from the server, we should move them to another path like /edx/var/<mfe_name>/static/

I also added configurations that would allow us to override the footer and the header for branding purposes. But I still need to test that.

So far, I used the native.sh script to deploy this in a server, and it is in:

https://mfe.stage.edunext.co
https://profile.mfe.stage.edunext.co
https://account.mfe.stage.edunext.co
https://gradebook.mfe.stage.edunext.co

The default users from a native installation should be available
(staff@example.com and honor@example.com with pass: edx).

Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).
  - [ ] Think about how this change will affect Open edX operators.  Have you updated the wiki page for the next Open edX release?

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @morenol! I've created [OSPR-5001](https://openedx.atlassian.net/browse/OSPR-5001) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by morenol** [comment]
Hi @uzairr, I noticed that you have open [a PR](https://github.com/edx/configuration/pull/6023) to deploy the account MFE, could you check this?

I think that we can use things from your PR and from my PR.

**Comment by natabene** [comment]
@morenol I see this is a draft. Please let me know once it is ready for our review.

**Comment by morenol** [comment]
Hi @natabene I think that this is ready for review. I would like to have reviews from edX and from the community

**Comment by nedbat** [comment]
@morenol You could mention this on the "[Deploying MFEs in the community](https://discuss.openedx.org/t/deploying-mfes-in-the-community/2868/8)" Discourse thread to get more community attention on it.

**Comment by xitij2000** [comment]
@morenol This PR looks great!  I had a similar one in progress, which used a slightly different mechanism. 

While I went for almost exactly the same script for deploying an MFE, I went for a different way of using that script to deploy individual MFEs. In this PR there is a separate role for each MFE deployment, and each of those roles is listed in `openedx_native.yml`. What I did instead was to create a single role to deploy all MFEs instead. That role would read a configuration that would list all MFE parameters in one place as follows:

```yaml

MFES:
  - repo: ...github.com/open-craft/frontend-app-xyz
    url: xyz.site.opencraft.hosting
    node_version: 12
    overrides:
      - {npm-pacakge}    
    settings:
      MY_APP_SETTING: "{LMS_URL}/some-path"
```

This can still be populated by variables such as `SANDBOX_ENABLE_ACCOUNT` etc.  My reasoning for this was to allow the deployment of custom or third-party MFEs. I hope that makes sense?

File: playbooks/roles/mfe/templates/edx/app/nginx/sites-available/concerns/mfe-common.j2 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{% if MFE_MAX_WEBSERVER_UPLOAD %}
+  client_max_body_size {{ MFE_MAX_WEBSERVER_UPLOAD }}M;
+{% endif %}
+{% if MFE_NGINX_READ_TIMEOUT %}
+    proxy_read_timeout {{ MFE_NGINX_READ_TIMEOUT }};
+{% endif %}
+
+    add_header 'Access-Control-Allow-Origin' '*';
```
Comment (regisb) [inline playbooks/roles/mfe/templates/edx/app/nginx/sites-available/concerns/mfe-common.j2]: This seems overly generous. Why is this necessary?

Comment (morenol) [reply playbooks/roles/mfe/templates/edx/app/nginx/sites-available/concerns/mfe-common.j2]: Actually, I think that it is not necessary. I am going to remove it now

**Comment by morenol** [comment]
>  What I did instead was to create a single role to deploy all MFEs instead. 

@xitij2000 It makes a lot of sense, I removed the gradebook, profile and account roles and instead I changed that to just use the mfe role. I am going to check how to do the thing that you are suggesting about the deployment of multiple MFEs from a single configuration.

**Comment by nedbat** [comment]
@fredsmith @adzuci Thoughts?

**Review by fredsmith — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@morenol 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/6039&entry.1484655318=https://github.com/morenol&entry.752974735=2020-09-23+16:22&entry.1917517419=2020-09-28+14:44&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by morenol** [comment]
These were the extra configurations used to configure https://mfe.stage.edunext.co

```yaml
EDXAPP_LMS_BASE: "mfe.stage.edunext.co"
EDXAPP_CMS_BASE: "studio.mfe.stage.edunext.co"

MFE_HOSTNAME: '~^((stage|prod)-)?{{ MFE_NAME }}.{{ EDXAPP_LMS_BASE }}'
EDXAPP_LMS_BASE_SCHEME: https
MFE_BASE_SCHEMA: https

NGINX_ENABLE_SSL: yes

NGINX_SSL_CERTIFICATE: 'mfe.pem'
NGINX_SSL_KEY: 'mfe.key'

EDXAPP_SESSION_COOKIE_DOMAIN: ".mfe.stage.edunext.co"

EDXAPP_CSRF_COOKIE_SECURE: true
EDXAPP_SESSION_COOKIE_SECURE: true 
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".mfe.stage.edunext.co"
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

EDXAPP_CORS_ORIGIN_WHITELIST:
- studio.mfe.stage.edunext.co
- gradebook.mfe.stage.edunext.co
- account.mfe.stage.edunext.co
- profile.mfe.stage.edunext.co

EDXAPP_CSRF_TRUSTED_ORIGINS:
- gradebook.mfe.stage.edunext.co
- account.mfe.stage.edunext.co
- profile.mfe.stage.edunext.co

# Needed to link the LMS instructor dashboard to the writable gradebook micro-frontend
EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: https://gradebook.mfe.stage.edunext.co

# Needed to link to the new profile micro-frontend.
EDXAPP_PROFILE_MICROFRONTEND_URL: https://profile.mfe.stage.edunext.co/u/

# Needed to link to the new account micro-frontend.
EDXAPP_ACCOUNT_MICROFRONTEND_URL: https://account.mfe.stage.edunext.co

EDXAPP_LMS_SSL_NGINX_PORT: 443
EDXAPP_CMS_SSL_NGINX_PORT: 443

module_lang: en_US.UTF-8
module_set_locale: True

# Added to override the value of ENABLE_CORS_HEADERS and ENABLE_CROSS_DOMAIN_CSRF_COOKIE
# Once this gets merged: https://github.com/edx/configuration/pull/5992/files
# We can just use EDXAPP_ENABLE_CORS_HEADERS and ENABLE_CROSS_DOMAIN_CSRF_COOKIE
EDXAPP_FEATURES:
    AUTH_USE_OPENID_PROVIDER: true
    AUTOMATIC_AUTH_FOR_TESTING: false
    CUSTOM_COURSES_EDX: false
    ENABLE_BULK_ENROLLMENT_VIEW: false
    ENABLE_COMBINED_LOGIN_REGISTRATION: true
    ENABLE_CORS_HEADERS: true
    ENABLE_COUNTRY_ACCESS: false
    ENABLE_CREDIT_API: false
    ENABLE_CREDIT_ELIGIBILITY: false
    ENABLE_CROSS_DOMAIN_CSRF_COOKIE: true
    ENABLE_CSMH_EXTENDED: true
    ENABLE_DISCUSSION_HOME_PANEL: true
    ENABLE_DISCUSSION_SERVICE: true
    ENABLE_EDXNOTES: false
    ENABLE_ENROLLMENT_RESET: false
    ENABLE_EXPORT_GIT: false
    ENABLE_GRADE_DOWNLOADS: true
    ENABLE_INSTRUCTOR_ANALYTICS: false
    ENABLE_MKTG_SITE: false
    ENABLE_MOBILE_REST_API: false
    ENABLE_OAUTH2_PROVIDER: false
    ENABLE_PUBLISHER: false
    ENABLE_READING_FROM_MULTIPLE_HISTORY_TABLES: true
    ENABLE_SPECIAL_EXAMS: false
    ENABLE_SYSADMIN_DASHBOARD: false
    ENABLE_THIRD_PARTY_AUTH: true
    ENABLE_VIDEO_UPLOAD_PIPELINE: false
    PREVIEW_LMS_BASE: mfe.stage.edunext.co
    SHOW_FOOTER_LANGUAGE_SELECTOR: false
    SHOW_HEADER_LANGUAGE_SELECTOR: false

# DISABLE unrelated services
SANDBOX_ENABLE_ANALYTICS_API: false
SANDBOX_ENABLE_INSIGHTS: false
SANDBOX_ENABLE_DISCOVERY: false
SANDBOX_ENABLE_ECOMMERCE: false
SANDBOX_ENABLE_NOTES: false
COMMON_ENABLE_SPLUNKFORWARDER: false
COMMON_ENABLE_DATADOG: false
```

Other than that, in order to have the MFE enabled in the LMS it is necessary to enable them through the waffle flags (account.redirect_to_microfrontend, learner_profile.redirect_to_microfrontend and grades.writable_gradebook). And also enable them in the site configuration with: 

```json
{
     "ENABLE_PROFILE_MICROFRONTEND": true,
    "ENABLE_ACCOUNT_MICROFRONTEND": true
}
```

I ran this with OPENEDX_RELEASE=master

**Comment by nedbat** [comment]
How do we come to an agreement about whether subdomains is the right solution for a simple out-of-the-box installation? The [BTR-14](https://openedx.atlassian.net/browse/BTR-14) Jira issue is about this.

**Comment by regisb** [comment]
> How do we come to an agreement about whether subdomains is the right solution for a simple out-of-the-box installation?

I think we all agree that subdomains are not the perfect solution, either for the native installation and for edx.org, and that we want to move to subpaths.

**Comment by morenol** [comment]
Should we move to subpaths before the release of Koa? 

I think that it is feasible, but in that case we would need to decide how that would be accomplished. It is clear to me which changes are needed in the frontend-platform and frontend-build to support that, but I am not sure yet how we should implement the changes to route the requests to the MFEs. We could just add a location block for each MFE in the lms nginx config, but that would not cover the case in which the main domain is not the lms and instead it is a MKTG site, for that case I guess that we would need a new layer to route the requests, but I guess that it is not in the scope of the native installation.

**Comment by regisb** [comment]
Can we please move the conversation to https://openedx.atlassian.net/jira/software/projects/BTR/boards/641?selectedIssue=BTR-14 ?
